### PR TITLE
Line comments

### PR DESCRIPTION
**Pull request by eanker**
After splitting a hunk, it is now possbile to add a comment in between.

### CODE REVIEW

File: bower.json [hunk 1]
```diff
@@ -21,7 +21,8 @@
     "tests"
   ],
   "dependencies": {
-    "polymer": "^1.8.0"
+    "polymer": "^1.8.0",
+    "paper-input": "PolymerElements/paper-input#^1.1.23"
```
Comment (txsmith) [inline bower.json]: `paper-button` and `paper-textarea` are still missing

File: code-highlighter.html [hunk 2]
```diff
@@ -225,18 +225,21 @@
       _splitHunk: function (e) {
         var path = this._getComposedPath(e)[0];
```
Comment (txsmith) [inline code-highlighter.html]: I think you missed this in the previous review: this line is equivalent to `var path = e.target`.

File: code-highlighter.html [hunk 3]
```diff
@@ -225,18 +225,21 @@
...
+            this.fire('add-hunk', newElem);
+          } else {
+            this.fire('add-comment');
+          }
         }
       },
-
       // For browsers that support neither Event.path and Event.composedPath.
       // from: https://stackoverflow.com/questions/36845515/mouseevent-path-equivalent-in-firefox-safari
       _getComposedPath: function (e) {
```
Comment (txsmith) [inline code-highlighter.html]: Using `e.target`, this method becomes unnecessary.

File: comment-container.html [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+      :host {
+        display: block;
+      }
+      
+      [hidden] {
+        display: none !important;
+      }
+    </style>
+    
+    <add-comment hidden$="[[hideAdd]]" hide="{{hideAdd}}" comments="{{comments}}"></add-comment>
```
Comment (txsmith) [inline comment-container.html]: This still needs an HTML import

Comment (txsmith) [reply comment-container.html]: `hidden$="[[hideAdd]]" hide="{{hideAdd}}"` -> `hidden$="{{hideAdd}}""

File: add-comment.html [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+        hide: {
+          type: Boolean,
+          notify: true
+        }
+      },
+      _addComment: function () {
+        this.push('comments', this.comment);
+        this.fire('toggle-add-comment');
+        this.comment = '';
+        this.hide = true;
```
Comment (txsmith) [inline add-comment.html]: `this.hidden = true`

File: add-comment.html [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+      },
+
+      _hide: function () {
+        this.fire('toggle-add-comment');
+        this.hide = true;
```
Comment (txsmith) [inline add-comment.html]: `this.hidden = true`

File: code-with-comment.html [hunk 7]
```diff
@@ -36,7 +45,25 @@
...
+        this.hideAddComments = false;
+        e.stopPropagation();
+        var elem = e.detail;
+        elem.comments = this.comments;
+        this.fire('add-hunk-with-comment', elem);
+        this.comments = [];
+      },
+      _showAddComment: function () {
+        this.fire('toggle-add-comment');
+        this.hideAddComments = false;
```
Comment (txsmith) [inline code-with-comment.html]: In the event handler above, you call `e.stopPropagation()` but here you do not. Why?

File: comment-container.html [hunk 8]
```diff
@@ -0,0 +1,36 @@
+<link rel="import" href="add-comment.html">
+<link rel="import" href="comment-list.html">
+<link rel="import" href="../paper-button/paper-button.html">
```
Comment (txsmith) [inline comment-container.html]: Unnecessary import

File: code-highlighter.html [hunk 9]
```diff
@@ -225,18 +225,21 @@
...
+        if (path.localName === 'iron-icon') {
+          if (lineNumber !== this._amountLines) {
+            var newElem = {};
+            var x = this.code.split('\n');
+            newElem.code = x.slice(lineNumber, this.code.length - 1).join('\n');
+            newElem.beforeStart = path.getAttribute('before');
+            newElem.afterStart = path.getAttribute('after');
+            newElem.index = this.index;
+            this.code = x.slice(0, lineNumber).join('\n');
+            this.fire('add-hunk', newElem);
```
Comment (txsmith) [inline code-highlighter.html]: I'd like to propose renaming this to `split-code` since this event does not add a hunk at all. It just splits the highlight in two. When adding comments, the amount of hunks remain the same.

Comment (txsmith) [reply code-highlighter.html]: Or `click-line`. This is more general and leaves handling of the 'last line clicked' case to the parent element. So then `code-with-comment` would decide that a comment has to be added when the last line is clicked.

File: test/code-highlighter.html [hunk 10]
```diff
@@ -85,15 +85,33 @@
...
+        });
+        elem.addEventListener('add-comment', function () {
+          assert.equal(countComment, 0);
+          countComment++;
+        });
+
+        MockInteractions.tap(icon);
+        MockInteractions.tap(icon);
+
+      });
```
Comment (txsmith) [inline test/code-highlighter.html]: This test does not work as intended. It finished before the two event handlers get the chance to do the assertions. Use `done` to prevent this.

File: test/code-highlighter.html [hunk 11]
```diff
@@ -85,15 +85,33 @@
...
         var icon = Polymer.dom(elem.root).querySelector('iron-icon');
-        flush(function () {
-          elem.addEventListener('add-hunk', function () {
-            done(new Error('This event should not have been fired'));
-          });
+        elem.addEventListener('add-hunk', function () {
+          done(new Error('This event should not have been fired'));
         });
         MockInteractions.tap(icon);
         setTimeout(done, 10);
```
Comment (txsmith) [inline test/code-highlighter.html]: 10 millisec timeout is not really enough. Slow builds will essentially disable this test because the timeout fires before any event gets dispatched. I'd suggest using 100 ms instead.

**Review by txsmith — CHANGES_REQUESTED** [review]

File: add-comment.html [hunk 12]
```diff
@@ -0,0 +1,49 @@
...
+          type: Array,
+          notify: true
+        },
+        hide: {
+          type: Boolean,
+          notify: true
+        }
+      },
+      _addComment: function () {
+        this.push('comments', this.comment);
```
Comment (txsmith) [inline add-comment.html]: It's possible to add an empty comment

**Review by txsmith — CHANGES_REQUESTED** [review]

File: bower.json (hunk 1 above)
Comment (eanker) [reply bower.json]: This is the dependency for `paper-textarea`:` paper-input/paper-textarea.html`
Will add paper-button and iron-icon too.

Comment (txsmith) [inline bower.json]: :+1:

File: code-highlighter.html (hunk 9 above)
Comment (eanker) [reply code-highlighter.html]: should we then also change `this.fire('add-comment')`, so these to lines fire the same event.
Or is this one fine.
`

File: comment-container.html (hunk 4 above)
Comment (eanker) [reply comment-container.html]: This doesn't work

File: code-highlighter.html (hunk 9 above)
Comment (txsmith) [reply code-highlighter.html]: Yes, code-highlighter should fire one event

File: test/code-highlighter.html [hunk 13]
```diff
@@ -73,7 +73,7 @@
       [0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13].forEach(function (count) {
```
Comment (txsmith) [inline test/code-highlighter.html]: Why not just a for-loop from 0 to 13?

File: test/code-highlighter.html [hunk 14]
```diff
@@ -41,7 +41,7 @@
       });
 
       it('tap button fires event', function () {
-        elem.addEventListener('add-hunk', function (e) {
+        elem.addEventListener('split-code', function (e) {
           assert.equal(e.detail.beforeStart, elem.beforeStart + 1)
           assert.equal(e.detail.afterStart, elem.afterStart + 1)
           assert.equal(e.detail.code, ' second line \n');
```
Comment (txsmith) [inline test/code-highlighter.html]: You forgot to call done() in some tests, which makes them always pass

File: test/code-highlighter.html [hunk 15]
```diff
@@ -52,17 +52,17 @@
...
-        elem.addEventListener('add-hunk', function (e) {
+        elem.addEventListener('split-code', function (e) {
           assert.equal(e.detail.beforeStart, 0)
           assert.equal(e.detail.afterStart, 1)
           assert.equal(e.detail.code, ' second line \n');
           assert.equal(e.detail.index, 1);
         });
-        elem.fire('add-hunk', {beforeStart: 0, afterStart: 1, code: ' second line \n', index: 1});
+        elem.fire('split-code', { beforeStart: 0, afterStart: 1, code: ' second line \n', index: 1 });
       });
```
Comment (txsmith) [inline test/code-highlighter.html]: This test is not testing anything. You fire the expected event manually.

File: test/code-highlighter.html [hunk 16]
```diff
@@ -85,20 +85,42 @@
       it('tapping on last line should not fire', function (done) {
```
Comment (txsmith) [inline test/code-highlighter.html]: It should...

File: test/code-highlighter.html [hunk 17]
```diff
@@ -85,20 +85,42 @@
...
         elem.code = 'one liner \n'
         var icon = Polymer.dom(elem.root).querySelector('iron-icon');
+        elem.addEventListener('split-code', function () {
+          done(new Error('This event should not have been fired'));
+        });
+        MockInteractions.tap(icon);
+        setTimeout(done, 100);
+      });
+
+      it('tapping twice only fires comment', function () {
```
Comment (txsmith) [inline test/code-highlighter.html]: This test is not valid in `code-highlighter` anymore. Maybe in `code-with-comments`?

File: code-highlighter.html [hunk 18]
```diff
@@ -223,40 +223,24 @@
...
-          while (currentElem) {
-            path.push(currentElem);
-            currentElem = currentElem.parentElement;
+          if (lineNumber !== this._amountLines) {
+            var x = this.code.split('\n');
+            newElem.code = x.slice(lineNumber, this.code.length - 1).join('\n');
+            newElem.beforeStart = path.getAttribute('before');
+            newElem.afterStart = path.getAttribute('after');
+            newElem.index = this.index;
+            this.code = x.slice(0, lineNumber).join('\n');
```
Comment (txsmith) [inline code-highlighter.html]: I think we want to alter the element's code in the parent element. IMO clicking a line should just fire an event and let the user of this element decide what to do with the event. This way, `code-highlighter` can be used in more scenario's.

File: code-highlighter.html [hunk 19]
```diff
@@ -223,40 +223,24 @@
...
+            this.fire('split-code', newElem);
```
Comment (txsmith) [inline code-highlighter.html]: The above comment would mean that this event would become `click-line` or something similar.

File: code-with-comment.html [hunk 20]
```diff
@@ -36,7 +44,21 @@
...
+        hideComments: Boolean
+      },
+      _splitCode: function (e) {
+        this.hideAddComments = false;
+        e.stopPropagation();
+        var elem = e.detail;
+        elem.comments = this.comments;
+        this.fire('add-hunk-with-comment', elem);
+        this.comments = [];
       }
```
Comment (txsmith) [inline code-with-comment.html]: Given my comment above, this element will become responsible for:
 - checking the click on the last line
 - editing the `code` property
 - firing the `add-code-with-comment` event to the parent

File: hunk-element.html [hunk 21]
```diff
@@ -25,29 +25,33 @@
...
   <script>
     Polymer({
       is: 'hunk-element',
 
       listeners: {
-        'add-hunk': '_addHunk'
+        'add-hunk-with-comment': '_addHunk'
       },
       properties: {
         hunks: {
```
Comment (txsmith) [inline hunk-element.html]: Can you rename this? It's an array of code chunks that together form one hunk.

File: test/add-comment.html [hunk 22]
```diff
@@ -0,0 +1,79 @@
...
+        assert.deepEqual(elem.comments, ['hello']);
+      });
+
+      it('Adding empty comment does not add', function () {
+        var addButton = Polymer.dom(elem.root).querySelector('#add');
+        elem.comment = '';
+        assert.deepEqual(elem.comments, []);
+        MockInteractions.tap(addButton);
+        assert.deepEqual(elem.comments, []);
+      });
```
Comment (txsmith) [inline test/add-comment.html:53]: You can also add a test for comments that only contain whitespace/tabs/etc

**Review by txsmith — CHANGES_REQUESTED** [review]

**Comment by eanker** [comment]
@txsmith could you look into this today?

**Review by txsmith — APPROVED** [review]
Confirmed working!
